File: packages/ocp-crypto/test/transcript.test.ts
import { describe, expect, it } from "vitest";

import {
  Transcript,
  bytesToHex,
  scalarToBytes,
} from "../src/index.js";

describe("Transcript", () => {
  // Backward-compatibility golden vector. Captured against the pre-fold
  // implementation; must remain unchanged after the fold is introduced
  // because single-challenge callers observe the same challenge digest.
  const SINGLE_CHALLENGE_GOLDEN =
    "a2e9007cc0b1a8acf959efd4e0a810bd90bc64b16d9b96b266c5f3b328960d05";

  it("single-challenge golden vector is unchanged (backward compat)", () => {
    const tr = new Transcript("ocp/v1/test");
    tr.appendMessage("msg1", new Uint8Array([1, 2, 3, 4]));
    const e = tr.challengeScalar("e");
    expect(bytesToHex(scalarToBytes(e))).toBe(SINGLE_CHALLENGE_GOLDEN);
  });

  it("two challenges on the same transcript differ (fold is in effect)", () => {
    const tr = new Transcript("ocp/v1/test");
    tr.appendMessage("msg1", new Uint8Array([1, 2, 3, 4]));
    const e1 = tr.challengeScalar("e");
    const e2 = tr.challengeScalar("e");
    expect(bytesToHex(scalarToBytes(e1))).not.toBe(bytesToHex(scalarToBytes(e2)));
  });

  it("first challenge equals the single-challenge golden even when a second is drawn", () => {
    // Belt-and-suspenders: the fold must only affect subsequent challenges,
    // never the first one.
    const tr = new Transcript("ocp/v1/test");
    tr.appendMessage("msg1", new Uint8Array([1, 2, 3, 4]));
    const e1 = tr.challengeScalar("e");
    // Drawing a second challenge must not retroactively change e1.
    tr.challengeScalar("e");
    expect(bytesToHex(scalarToBytes(e1))).toBe(SINGLE_CHALLENGE_GOLDEN);
  });

  it("second challenge binds earlier appended messages (binding)", () => {
    // Two transcripts with identical prefix through the first challenge,
    // but differing in the message appended between challenge 1 and
    // challenge 2, must yield different second challenges. Without the
    // fold they would match whenever the intervening append was the same
    // — which isn't what we're testing here — but the binding guarantee
    // is that *any* difference in the inter-challenge messages flows into
    // the second challenge.
    const trA = new Transcript("ocp/v1/test");
    trA.appendMessage("m", new Uint8Array([0xaa]));
    const e1A = trA.challengeScalar("e");
    trA.appendMessage("extra", new Uint8Array([0x01]));
    const e2A = trA.challengeScalar("e");

    const trB = new Transcript("ocp/v1/test");
    trB.appendMessage("m", new Uint8Array([0xaa]));
    const e1B = trB.challengeScalar("e");
    trB.appendMessage("extra", new Uint8Array([0x02]));
    const e2B = trB.challengeScalar("e");

    // First challenges are identical: same domain + same single append.
    expect(bytesToHex(scalarToBytes(e1A))).toBe(bytesToHex(scalarToBytes(e1B)));
    // Second challenges differ because the intervening append differs.
    expect(bytesToHex(scalarToBytes(e2A))).not.toBe(bytesToHex(scalarToBytes(e2B)));
  });

  it("second challenge binds the first challenge's label (fold tags label)", () => {
    // Two transcripts identical except for the label of the first
    // challenge. The first challenges differ (labels are part of the
    // challenge hash). The second challenges must also differ — the fold
    // binds the first-challenge label via the "chal" tag.
    const trA = new Transcript("ocp/v1/test");
    trA.appendMessage("m", new Uint8Array([0xaa]));
    trA.challengeScalar("e");
    const e2A = trA.challengeScalar("e");

    const trB = new Transcript("ocp/v1/test");
    trB.appendMessage("m", new Uint8Array([0xaa]));
    trB.challengeScalar("f");
    const e2B = trB.challengeScalar("e");

    expect(bytesToHex(scalarToBytes(e2A))).not.toBe(bytesToHex(scalarToBytes(e2B)));
  });
});

File: apps/cosmos/internal/ocpcrypto/transcript_test.go
package ocpcrypto

import (
	"bytes"
	"testing"
)

// Backward-compatibility golden vector. Captured against the pre-fold
// implementation in both TS and Go; must remain unchanged after the fold
// is introduced because single-challenge callers see the same challenge
// digest. The TS and Go Transcript implementations are required to be
// bit-for-bit interoperable — this value is the same on both sides.
const singleChallengeGoldenHex = "a2e9007cc0b1a8acf959efd4e0a810bd90bc64b16d9b96b266c5f3b328960d05"

func TestTranscript_SingleChallengeGolden(t *testing.T) {
	// Backward-compat: one appendMessage + one challengeScalar must yield
	// exactly the pre-fold golden value.
	tr := NewTranscript("ocp/v1/test")
	if err := tr.AppendMessage("msg1", []byte{1, 2, 3, 4}); err != nil {
		t.Fatalf("append: %v", err)
	}
	e, err := tr.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("challenge: %v", err)
	}
	want, err := hexToBytes(singleChallengeGoldenHex)
	if err != nil {
		t.Fatalf("hex: %v", err)
	}
	if !bytes.Equal(e.Bytes(), want) {
		t.Fatalf("single-challenge golden mismatch: got=%s want=%s",
			bytesToHex(e.Bytes()), bytesToHex(want))
	}
}

func TestTranscript_TwoChallengesDiffer(t *testing.T) {
	// Multi-challenge: the second challenge must differ from the first,
	// proving that the fold is actually binding subsequent challenges.
	tr := NewTranscript("ocp/v1/test")
	if err := tr.AppendMessage("msg1", []byte{1, 2, 3, 4}); err != nil {
		t.Fatalf("append: %v", err)
	}
	e1, err := tr.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("challenge1: %v", err)
	}
	e2, err := tr.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("challenge2: %v", err)
	}
	if bytes.Equal(e1.Bytes(), e2.Bytes()) {
		t.Fatalf("expected second challenge to differ from first; got %s for both",
			bytesToHex(e1.Bytes()))
	}
}

func TestTranscript_FirstChallengeUnchangedByLaterDraws(t *testing.T) {
	// The fold must not retroactively affect the first challenge.
	tr := NewTranscript("ocp/v1/test")
	if err := tr.AppendMessage("msg1", []byte{1, 2, 3, 4}); err != nil {
		t.Fatalf("append: %v", err)
	}
	e1, err := tr.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("challenge1: %v", err)
	}
	// Draw a second challenge and confirm e1's bytes are unchanged.
	if _, err := tr.ChallengeScalar("e"); err != nil {
		t.Fatalf("challenge2: %v", err)
	}
	want, err := hexToBytes(singleChallengeGoldenHex)
	if err != nil {
		t.Fatalf("hex: %v", err)
	}
	if !bytes.Equal(e1.Bytes(), want) {
		t.Fatalf("first challenge mutated by second draw: got=%s want=%s",
			bytesToHex(e1.Bytes()), bytesToHex(want))
	}
}

func TestTranscript_SecondChallengeBindsInterveningMessage(t *testing.T) {
	// Two transcripts with the same first challenge but a different
	// inter-challenge message must produce different second challenges.
	trA := NewTranscript("ocp/v1/test")
	_ = trA.AppendMessage("m", []byte{0xaa})
	e1A, err := trA.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("A.e1: %v", err)
	}
	_ = trA.AppendMessage("extra", []byte{0x01})
	e2A, err := trA.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("A.e2: %v", err)
	}

	trB := NewTranscript("ocp/v1/test")
	_ = trB.AppendMessage("m", []byte{0xaa})
	e1B, err := trB.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("B.e1: %v", err)
	}
	_ = trB.AppendMessage("extra", []byte{0x02})
	e2B, err := trB.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("B.e2: %v", err)
	}

	if !bytes.Equal(e1A.Bytes(), e1B.Bytes()) {
		t.Fatalf("expected identical first challenges, got A=%s B=%s",
			bytesToHex(e1A.Bytes()), bytesToHex(e1B.Bytes()))
	}
	if bytes.Equal(e2A.Bytes(), e2B.Bytes()) {
		t.Fatalf("expected differing second challenges; both were %s",
			bytesToHex(e2A.Bytes()))
	}
}

func TestTranscript_SecondChallengeBindsFirstChallengeLabel(t *testing.T) {
	// Two transcripts identical except for the label used for the first
	// challenge. The first challenges differ (label flows into the
	// challenge hash). The second challenges must also differ, proving
	// that the fold tags the challenge with its label.
	trA := NewTranscript("ocp/v1/test")
	_ = trA.AppendMessage("m", []byte{0xaa})
	if _, err := trA.ChallengeScalar("e"); err != nil {
		t.Fatalf("A.e: %v", err)
	}
	e2A, err := trA.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("A.e2: %v", err)
	}

	trB := NewTranscript("ocp/v1/test")
	_ = trB.AppendMessage("m", []byte{0xaa})
	if _, err := trB.ChallengeScalar("f"); err != nil {
		t.Fatalf("B.f: %v", err)
	}
	e2B, err := trB.ChallengeScalar("e")
	if err != nil {
		t.Fatalf("B.e2: %v", err)
	}

	if bytes.Equal(e2A.Bytes(), e2B.Bytes()) {
		t.Fatalf("expected differing second challenges when first labels differ; both were %s",
			bytesToHex(e2A.Bytes()))
	}
}
